---
title: "Haplotype calling pipeline"
author: "Kelsey Sumner"
date: "April 22, 2021"
output: 
  html_document:
    toc: true
    number_sections: false
    toc_float:
      collapsed: false
      smooth_scroll: true
    toc_depth: 3
    theme: lumen
    highlight: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Last Updated `r format(Sys.Date(), format="%m/%d/%Y")`


**Approach:** 
Using human or mosquito samples from Webuye, Kenya, evaluate the extent to which there are unique haplotypes among two polymorphic gene targets: AMA1, CSP.

**Data aggregation:** 
Method: Targeted amplicon deep sequencing which produces forward and reverse fastq files for each sample.


# **Sequence Processing & Cleaning in Terminal**

## BBmap Protocol 
* Describe the overall process here
    + Notes: 
        1. *write important notes*
        2. *write notes*
    + Code for running bbmap: `script.perl`
    + Input data for running bbmap path: `path/folder`
    + Output data from bbmap path: `path/folder`
    
## Any other subsection of processing


# **Haplotype Inference & Data Set Creation [Kelsey]**

## Running DADA2 to Create Haplotypes
* Create haplotypes from the cleaned, demultiplexed, and preprocessed targeted amplicon sequence data using DADA2 version 1.8, following this tutorial: https://benjjneb.github.io/dada2/tutorial.html.
* Run DADA2 for haplotype inference on the data output from the bbmap preprocessing in terminal
    + Code for running DADA2 on Github: `taylorlab/MESA_project/code/haplotype_inference/MESA_dada2.R`
    + Input data at Duke Box: 
        - 514 samples for each target (3 targets - AMA, CSP, HistB)
        - For AMA: `Webuye MESA Sequence Data > Mapped Cut Reads > AMA_haplotypes > AMA > input_fastqs`
        - For CSP: `Webuye MESA Sequence Data > Mapped Cut Reads > CSP_haplotypes > input_fastqs`
        - For HistB: `Webuye MESA Sequence Data > Mapped Cut Reads > HistB_haplotypes > input_fastqs`
    + Output data at Duke Box: 
        - For AMA: `Webuye MESA Sequence Data > Mapped Cut Reads > AMA_haplotypes > AMA > Original Results`
        - For CSP: `Webuye MESA Sequence Data > Mapped Cut Reads > CSP_haplotypes > Original Results`
        - For HistB: `Webuye MESA Sequence Data > Mapped Cut Reads > HistB_haplotypes > Original Results`
    + Notes: 
        1. *All of the haplotype inference is done in R*
        2. *The processing is set up to mulithread the samples through multiple cores of a Mac (for increased efficiency) so needs to be run on a Mac.*
        3. *Can be run locally on a computer.*
        4. *Chimeras were removed in this step using DADA2.*

## Cleaning Haplotype Output
* Clean the DADA2 haplotype inference output to only include samples and haplotypes that occurred in a high enough prevalence
    + Inclusion criteria for the samples and haplotypes: 
        1. *Kept only DBS human samples - removed 3D7 DBS control, 3D7 culture control, and 7g83D7Dd2*
        2. *Kept haplotypes that have >50 reads and are present in greater than 1% of the data*
    + Code for cleaning initial haplotype output: `taylorlab/MESA_project/code/haplotype_inference/MESA_haplotypeanalysis_nocontrols.R`
* Merge sample information to create one final cleaned data set of the processing from the initial fastq files to the final, cleaned haplotype data set
    + First run this code to create a cleaned trackReadsThroughPipeline file and finalizing haplotype output: `taylorlab/MESA_project/code/haplotype_inference/MESA_haplotypeoverlap.R`
    + Second run this code for creating merged sample information: `taylorlab/MESA_project/code/haplotype_inference/MESA_merge_sample_info.R`
* The data for both these steps was outputted to the same folder:
    + Input data: locations indicated in the R script (multiple files)
    + Output data at Duke Box: 
        - For AMA: `Deidentified data for Taylor lab > clean_files_for_lab > 23AUG2018 Haplotype Results > 23AUG2018 AMA MESA Update`
        - For CSP: `Deidentified data for Taylor lab > clean_files_for_lab > 23AUG2018 Haplotype Results > 23AUG2018 CSP MESA Update`
        - For HistB: `Deidentified data for Taylor lab > clean_files_for_lab > 23AUG2018 Haplotype Results > 23AUG2018 HistB MESA Update`
    + Notes: 
        1. *The data cleaning process had many steps and interim data sets (all found within the XXX_haplotypes folder for each target)*

## Combining Haplotype Calling, qPCR, & Social Demographic Data
* Clean the social demographic (meta data) data set. 
    + Code for cleaning initial social demographic data set: `taylorlab/MESA_project/code/haplotype_inference/MESA_clean_meta_data.R`
    + Cleaned social demographic data set (before the master merge with everything else): `Deidentified data for Taylor lab > clean_files_for_lab > mesa_data_clean.csv`
* Merge the haplotype calling results with the social demographic data set:
    + Code for merging the haplotype, social demographic data sets, qpcr, and lab key data sets: `taylorlab/MESA_project/code/haplotype_inference/MESA_finalstep_mergeall3datasets.R`
    + Output data set (final, clean, merged data set) for future analyses at Duke Box: `Deidentified data for Taylor lab > clean_files_for_lab > MESA_merged_final.csv`
    + Clean lab key (used to connect all inventories if needed): `Deidentified data for Taylor lab > clean_files_for_lab > clean_lab_key.csv`
    + Notes:
    1. *Many input data sets were used which are all specified at the top of the R scripts for the specific cleaning processes.*
    2. *Duplicates that occurred on the same plate were removed from the data set. Duplicates that occurred on different plates were renamed with a unique ID.*


# **Haplotype Analyses [Kelsey & Cody]**

## Haplotype Sharing [Kelsey]
* Create one combined data set that is an edgelist of pairwise comparisons of samples. The edgelist shows the proportion of shared haplotypes, number of haplotypes shared, and total number of unique haplotypes between each pair of samples for each target (AMA, CSP, HistB). Note: this is calculating haplotype sharing at the individual-level.
    + Code for calculating haplotype sharing between pairs (at the individual-level): `taylorlab/MESA_project/code/haplotype_sharing/MESA_pairwise_comparisons.R`
    + Data for combined haplotype sharing: 
`Webuye MESA Sequence Data > Mapped Cut Reads > all_targets`
* Calculate the amount of haplotype sharing at the village level for the CSP haplotype (used this for a visualization for the 2018 ASMTH conference).
    + Code for calculating haplotype sharing between households (at the village level): `taylorlab/MESA_project/code/haplotype_sharing/MESA_summarize_haplotype_sharing_HHlevel.R`


## Prevalence CSP 3D7 Haplotype [Kelsey]
* Calculate the prevalence of the the CSP 3D7 haplotype among all the samples that had haplotypes in the CSP data set. Haplotype 3 was found to be the CSP 3D7 haplotype in the haplotype data set after haplotype sequences were aligned using an online multialignment software (Clustal Omega: https://www.ebi.ac.uk/Tools/msa/clustalo/)
    + Code for calculating prevalence CSP 3D7 haplotype: `taylorlab/MESA_project/code/CSP_3D7_prevalence/MESA_prev_3D7_CSP_haplotype.R`
* Create shapefiles of the prevalence of the CSP 3D7 haplotype in each household to use to perform local and global spatial autocorrelation tests in GeoDa.
    + Code for creating shapefiles of CSP haplotype prevalence in each household: `taylorlab/MESA_project/code/CSP_3D7_prevalence/MESA_make_CSP_shapefile.R`
    + Shapefiles created of the prevalence of the CSP haplotype in each household: `Webuye MESA Sequence Data > Mapped Cut Reads > CSP_haplotypes > mesa_csp_shapefiles`
* Run a spatial regression model to assess if household prevalence of the 3D7 pfcsp haplotype is influenced by each household’s month of data collection, altitude, Anopheles mosquito density, and average distance to another household.


## Haplotype Visualizations [Kelsey]
* Created visualizations for the poster for the 2018 American Society of Tropical Medicine & Hygiene conference in New Orleans
    + Code for creating ASTMH 2018 poster visualizations: `taylorlab/MESA_project/code/visualizations/MESA_ASTMH_visualizations.R`
    + Figures for the poster are located at: `Webuye MESA Sequence Data > ASTMH > Poster > Figures`
* Created a visualization of when each haplotype was present over time with the x axis being each month during the cross-sectional study and the y-axis being every haplotype. 
    + Will just do this for both the AMA1 and CSP haplotypes
    + Code for creating haplotype visualization over time: `taylorlab/MESA_project/code/visualizations/MESA_haplotype_presence_over_time.R`
    

## Haplotype Phylogenetic Analyses and Population Genetics [Cody]
* More information to be included.
* Could combine Cody's haplotype phlyogenetic analyses and pouplation genetics work with Kelsey's descriptive and spatial haplotype work in a manuscript. 



# **Cross-sectional analyses [Kelsey]**

## False negative rapid diagnostic tests
* Standardize Ct values for each sample to calculate the parasites/uL in each sample
    + Code for standardizing qPCR curves: `taylorlab/MESA_project/code/crosssectional_analyses/MESA_standardize_qpcr_curves.R`
    + Notes:
    1. *Recalculated parasitemia values for each duplicate for each participant by creating a new standard curve with only standards 1-8 (concentrations from 1-2000 parasites/uL). Originally, standard curve based on standards 1-10 (0.1-2000 parasites/uL).*
    2. *Censoring criteria for parasitemia:*
        - *Recoding of CT values:*
            - *CT values of 0 recoded to missing (NA).*
            - *CT values of “undetermined” recoded to missing (NA) – however, these really signify negative results.*
            - *CT values 0 < CT < 40 for human beta-tubulin (Hb) are considered valid detection.*
            - *CT values of 0 < CT < 40 for Plasmodium falciparum are considered valid detection UNLESS only one replicate amplifies Pf, then the replicate that amplified has to have CT values of 0 < CT < 38 for Plasmodium falciparum to be considered valid.*
            - *Check that the standards (especially the first few) have amplified for human beta-tubulin and Plasmodium falciparum.*
        - *CT value censoring rules:*
            - *Pf positive via PCR:* 
                - *If 2 replicates amplified Pf: Pf CT values 0 < CT < 40*
                - *If 1 replicate amplified Pf: Pf CT values 0 < CT < 38*
            - *Pf negative via PCR: Pf CT undetermined (or Pf CT > 38 if one replicate), Hb CT values 0 < CT < 40*
            - *Pf missing via PCR: Pf CT values of 0 or sample not tested with PCR, Hb CT undetermined or 0 or sample not tested for PCR*
    3. *Rules for determining if participant is positive/negative/missing for Pf parasitemia:*
        - *If 1 replicate Pf positive per criteria above, used that replicate's standardized parasitemia for the participant to create combined parasitemia. Sample is positive for Pf in pf_pcr_infection_status variable.*
        - *If both replicates Pf positive per criteria above, averaged the replicates standardized parasitemia for the participant to create combined parasitemia. Sample is positive for Pf in pf_pcr_infection_status variable.*
        - *If both replicates Pf negative per criteria above, parasitemia is missing for combined parasitemia. Sample is negative for Pf in pf_pcr_infection_status variable.* 
        - *If both replicates missing via PCR per criteria above, parasitemia is missing for combined parasitemia. Sample is missing data for Pf in pf_pcr_infection_status variable.*  
* Assess the amount of rapid diagnostic tests (RDTs) that produced false negative results comparing the RDT results to the qPCR results with a negative RDT result and positive qPCR result indicating a false negative RDT result. 
* Compare the WHO lot testing reported pseudo-false negative rate for the rapid diagnostic test used (Carestart?) at 200 parasites/uL to the false negative rate observed in the cross-sectional data.
* Calculate the limit of detection for P. falciparum parasitemia for the cross-sectional study. 

